mlos_bench/mlos_bench/tests/launcher_test.py

```diff
@@ -37,27 +39,70 @@ def local_exec_service() -> LocalExecService:
     }))
 
 
-def test_launch_main_app(root_path: str,
-                         local_exec_service: LocalExecService) -> None:
+def _launch_main_app(root_path: str, local_exec_service: LocalExecService,
+                     cli_config: str, re_expected: List[str]) -> None:
     """
-    Run mlos_bench command-line application with mock config and check the results in the log.
+    Run mlos_bench command-line application with given config
+    and check the results in the log.
     """
     with local_exec_service.temp_dir_context() as temp_dir:
 
-        log_path = path_join(temp_dir, "mock-bench.log")
-        cmd = "./mlos_bench/mlos_bench/run.py" + \
-              " --config mlos_bench/mlos_bench/tests/config/cli/mock-bench.jsonc" + \
-              f" --log_file '{log_path}'"
-        (return_code, _stdout, _stderr) = local_exec_service.local_exec([cmd], cwd=root_path)
-
+        log_path = path_join(temp_dir, "mock-test.log")
+        (return_code, _stdout, _stderr) = local_exec_service.local_exec(
+            [f"./mlos_bench/mlos_bench/run.py {cli_config} --log_file '{log_path}'"],
+            cwd=root_path)
         assert return_code == 0
 
-        with open(log_path, "rt", encoding="utf-8") as fh_out:
-            best_score_lines = [
-                ln.strip() for ln in fh_out.readlines()
-                if " INFO Env: Mock environment best score: " in ln
-            ]
-            assert len([
-                ln for ln in best_score_lines
-                if " best score: 65.67" in ln
-            ]) == 1
+        try:
+            iter_expected = iter(re_expected)
+            re_log = re.compile(next(iter_expected))
+            with open(log_path, "rt", encoding="utf-8") as fh_out:
+                for ln in fh_out:
+                    if re_log.match(ln):
+                        re_log = re.compile(next(iter_expected))
```

Comment:
This opens and reads the file multiple times for each regex.

What about something more like the following:

```python
regexes = {pattern: re.compile(pattern) for pattern in patterns}
with open(...) as fh:
   for line in fh:
      for pattern in regexes:
          regex = regexes[pattern]
          if regex.match(line):
              del(regexes[pattern])
      if not regexes:
          break
if regexes:
    raise Exception("The following regexes were not matched: " + str(regexes))
```

Reply:
Nm, I missed the `next(iter_...` part *inside* the line loop.
Yours is actually even better than mine since you only need to evaluate a single regex (with expected order) for each line during the leading portion of the file.
Kudos!